lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,647 @@
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
+        1 => {
+            tx.execute_batch(include_str!("../../sql/migrations/0001.sql"))
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
+        2 => {
+            tx.execute_batch(include_str!("../../sql/migrations/0002.sql"))
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
+        _ => {
+            return Err(MigrationError::MigrationNumberDoesNotExist(
+                migration_number,
+                CURRENT_DATA_VERSION,
+            ));
+        }
+    }
+    tx.pragma_update(None, "user_version", &(migration_number + 1))
+        .map_err(|e| MigrationError::TransactionFailed(migration_number, format!("{}", e)))?;
+    tx.commit()
+        .map_err(|_| MigrationError::CommitFailed(migration_number))
+}
+
+/// Information about one of the user's keys
+#[derive(Debug)]
+pub struct PersonalKey {
+    /// Flag saying if the key will be used (there can only be one active key at a time)
+    pub active: bool,
+    /// The public key's tag. Used to identify the key pair
+    pub public_key_id: String,
+    /// The raw value of the public key in base64
+    pub public_key_value: String,
+    /// The location in the file system of the private key
+    pub private_key_location: Option<String>,
+    /// The type of private/public key this is
+    pub key_type_identifier: String,
+    /// The time at which the key was registered with wapm
+    pub date_created: OffsetDateTime,
+}
+
+fn get_active_personal_key(conn: &Connection) -> anyhow::Result<PersonalKey> {
+    let mut stmt = conn.prepare(
+        "SELECT active, public_key_value, private_key_location, date_added, key_type_identifier, public_key_id FROM personal_keys 
+         where active = 1",
+    )?;
+
+    let result = stmt
+        .query_map(params![], |row| {
+            Ok(PersonalKey {
+                active: row.get(0)?,
+                public_key_value: row.get(1)?,
+                private_key_location: row.get(2)?,
+                date_created: {
+                    use time::format_description::well_known::Rfc3339;
+                    let time_str: String = row.get(3)?;
+                    OffsetDateTime::parse(&time_str, &Rfc3339)
+                        .unwrap_or_else(|_| panic!("Failed to parse time string {}", &time_str))
+                },
+                key_type_identifier: row.get(4)?,
+                public_key_id: row.get(5)?,
+            })
+        })?
+        .next();
+
+    if let Some(res) = result {
+        Ok(res?)
+    } else {
+        Err(anyhow!("No active key found"))
+    }
+}
+
+mod interfaces {
```

Comment:
Should this be moved to its own file?

Reply:
I would like to keep it here, so that we have all of the publishing logic in one file (init.rs, login.rs, publish.rs). But it should be in a separate module so it doesn't interfere with anything else.